File: 151_160/MaximumProductSubarray.py
# 152. Maximum Product Subarray
# Dynamic Programming in O(n): 维持一个到当前元素的乘积最大值(max_curr)和最小值(min_curr)
# 状态转移方程:
# max_curr[i] = Max(A[i], max_curr[i-1] * A[i], min_curr * A[i])
# min_curr[i] = Min(A[i], max_curr[i-1] * A[i], min_curr * A[i])
class Solution:
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return None

        n = len(nums)
        maxPro = - float('Inf')
        maxCurr, minCurr = 1, 1
        for i in range(n):
            maxLast = maxCurr
            minLast = minCurr
            maxCurr = max(nums[i], maxLast*nums[i], minLast*nums[i])
            minCurr = min(nums[i], maxLast*nums[i], minLast*nums[i])
            print(nums[i], maxCurr, minCurr)
            maxPro = max(maxPro, maxCurr)

        return maxPro


print(Solution().maxProduct([-4, -3, -2]))

File: go_solution/151_160/maximumProductSubarray/maximumProductSubarray.go
package maximumProductSubarray

// 152. Maximum Product Subarray

func maxProduct(nums []int) int {
	if len(nums) == 1 {
		return nums[0]
	}
	max, min := nums[0], nums[0]
	rlt := max
	for i := 1; i < len(nums); i++ {
		if nums[i] > 0 {
			if nums[i]*max > nums[i] {
				max = nums[i] * max
			} else {
				max = nums[i]
			}
			if nums[i]*min < nums[i] {
				min = nums[i] * min
			} else {
				min = nums[i]
			}
		} else {
			tmp := max
			if nums[i]*min > nums[i] {
				max = nums[i] * min
			} else {
				max = nums[i]
			}
			if nums[i]*tmp < nums[i] {
				min = nums[i] * tmp
			} else {
				min = nums[i]
			}
		}
		if rlt < max {
			rlt = max
		}
	}
	return rlt
}
